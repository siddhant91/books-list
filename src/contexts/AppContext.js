import { useState, createContext } from 'react';
import { node } from 'prop-types';

const AppContext = createContext();

const AppProvider = ({ children }) => {
	const [loaderVisible, setLoaderVisible] = useState(false);

	return (
		<AppContext.Provider value={[loaderVisible, setLoaderVisible]}>{children}</AppContext.Provider>
	);
};

AppProvider.propTypes = {
	children: node.isRequired,
};
console.log(1);
setTimeout(() => {
	console.log(2);
}, 1000);

export { AppContext, AppProvider };
