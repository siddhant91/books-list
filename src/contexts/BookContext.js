import { useState, createContext } from 'react';
import { node } from 'prop-types';

const BookContext = createContext();

const BookProvider = ({ children }) => {
	const [booksList, setBooksList] = useState([]);

	return <BookContext.Provider value={[booksList, setBooksList]}>{children}</BookContext.Provider>;
};

BookProvider.propTypes = {
	children: node.isRequired,
};

export { BookContext, BookProvider };
