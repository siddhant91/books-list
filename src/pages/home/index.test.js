import { render, cleanup, waitFor, screen, act } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import Home from './index';
import { AppProvider } from '../../contexts/AppContext';
import { BookProvider } from '../../contexts/BookContext';
import mockBooksList from '../../../mockData/booksList';

afterEach(cleanup);

const mock = new MockAdapter(axios);

const renderComponent = (data = {}, mockError = false) => {
	act(() => {
		const mockData = data;
		if (mockError) {
			const mockErrorObj = {
				status: 400,
				errorMessage: 'bad Request',
			};
			mock.onGet().reply(400, mockErrorObj);
		} else {
			mock.onGet().reply(200, mockData);
		}
	});

	return (
		<AppProvider>
			<BookProvider>
				<Home />
			</BookProvider>
		</AppProvider>
	);
};

it('Matches snapshot', async () => {
	await act(async () => {
		const { asFragment } = render(renderComponent());
		expect(asFragment()).toMatchSnapshot();
	});
});

it('renders the title ', async () => {
	const { findByText } = render(renderComponent());
	expect(await findByText('Books')).toBeInTheDocument();
});

it('renders the title ', async () => {
	const { findByText } = render(renderComponent());
	expect(await findByText('Books')).toBeInTheDocument();
});

it('fetches the empty data ', async () => {
	render(renderComponent());
	expect(await screen.findByText('Sorry no books to Display')).toBeInTheDocument();
});

it('fetches the data ', async () => {
	render(renderComponent(mockBooksList));
	const bookCards = await screen.findAllByTestId('bokl-card');
	await waitFor(() => {
		expect(bookCards).toHaveLength(1);
	});
});

it('displays generic error and closes Modal', async () => {
	const { findByTestId } = render(renderComponent(mockBooksList, true));
	const errorTextElement = await findByTestId('generic-error');
	await waitFor(() => {
		expect(errorTextElement).toBeInTheDocument();
	});
});
