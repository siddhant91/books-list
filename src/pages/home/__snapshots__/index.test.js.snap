// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bokl-homepage"
  >
    <div
      class="bokl-page-layout"
    >
      <div
        class="bokl-page-layout__menu"
        data-testid="menu-container"
      >
        <button
          aria-label="open Menu"
          class="text-left p-0 d-flex align-items-center just-icon icon bokl-button-component"
          data-testid="toggle-menu"
          type="button"
        >
          <i
            aria-hidden="true"
            class="fa fa-bars"
          />
          <span
            class="bokl-page-layout__menu--item-title"
          >
            Menu
          </span>
        </button>
        <div
          class="d-flex flex-column mt-3"
          data-testid="menu-items"
        >
          <button
            aria-label="Content Management"
            class="text-left p-0 d-flex align-items-center mt-3 just-icon icon bokl-button-component"
            data-testid="menu-item-btn"
            type="button"
          >
            <i
              aria-hidden="true"
              class="fa fa-window-maximize"
            />
            <span
              class="bokl-page-layout__menu--item-title "
            >
              Content Management
              <i
                aria-hidden="true"
                class="fa fa-angle-down"
              />
            </span>
          </button>
          <button
            aria-label="Courses"
            class="text-left p-0 d-flex align-items-center mt-3 just-icon icon bokl-button-component"
            data-testid="menu-item-btn"
            type="button"
          >
            <i
              aria-hidden="true"
              class="fa fa-file"
            />
            <span
              class="bokl-page-layout__menu--item-title "
            >
              Courses
            </span>
          </button>
        </div>
      </div>
      <div
        class="bokl-page-layout__main"
      >
        <header
          class="bokl-page-layout__header d-flex align-items-center"
        >
          <img
            alt="Logo"
            class="bokl-page-layout__header--logo"
            data-testid="logo-img"
            src="Logo.png"
          />
        </header>
        <main
          class="bokl-page-layout__content"
        >
          <div
            class="bokl-homepage--content"
          >
            <div
              class="bokl-homepage--content__title d-flex flex-column flex-lg-row justify-content-between"
            >
              <h1>
                Books
              </h1>
              <button
                class="primary medium bokl-button-component"
                data-testid="create-book"
                type="button"
              >
                Create New Book
              </button>
            </div>
            <div
              class="bokl-homepage--content__main"
            >
              <div
                class="bokl-search-form"
              >
                <form
                  action="#"
                  autocomplete="off"
                  data-testid="searchBook-form"
                >
                  <div
                    class="bokl-text-input"
                  >
                    <label
                      class="d-flex bokl-text-input--label"
                      for="Search For Books"
                    >
                      <span>
                        <i
                          aria-hidden="true"
                          class="fa fa-search"
                        />
                      </span>
                      <input
                        aria-label="TextInput"
                        class=""
                        data-testid="TextInput"
                        id="Search For Books"
                        name="searchText"
                        placeholder="Search"
                        type="search"
                        value=""
                      />
                    </label>
                  </div>
                  <button
                    class="d-none"
                    type="submit"
                  >
                    Submit
                  </button>
                </form>
              </div>
              <div
                class="bokl-books-listing"
              >
                <h2
                  class="bokl-books-listing--title"
                >
                  All Books
                </h2>
                <div
                  class="bokl-books-listing--list mt-5 d-flex flex-wrap justify-content-between"
                >
                  <p>
                    Sorry no books to Display
                  </p>
                </div>
              </div>
            </div>
          </div>
        </main>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
